cpp/src/arrow/compute/expression_test.cc

```diff
@@ -895,8 +895,13 @@ TEST(Expression, ExecuteCallWithNoArguments) {
 
```

Comment:
Please rename this test case to `TEST(Expression, SimplifyImpureFunctionCall)`

Also, please add an impure function which takes arguments so that we can verify that arguments still get simplified:

```c++
  auto pure_expr = call("add", {field_ref("i32"), literal(3)});
  Simplify{
      call("my_seeded_random", {pure_expr})
  }
      .WithGuarantee(equal(field_ref("i32"), literal(1))
      .Expect(call("my_seeded_random", {literal(4)}));
```